File: notebook/loss.R
pacman::p_load(ggplot2, readr, dplyr, scales, crayon)

args <- commandArgs(trailingOnly = TRUE)
batch_skip = if(length(args) > 0) as.integer(args[1]) else 0

draw_plot <- function(last_scheduler_length=0){
  # Read the first three columns of the metrics file
  data <- read_tsv('metrics/multitask_loss.tsv', col_names = c('type', 'batch', 'loss', 'lr'), skip =1, show_col_types = FALSE) 
  data <- data |> mutate(type = ifelse(type=="scheduler","sched",type))
  scheduler_length = length(data |> filter(type == "sched") |> pull(loss))
  if(scheduler_length == last_scheduler_length){
    return(last_scheduler_length)
  }
  data <- data |> mutate(batch = row_number()) 
  data <- data |> filter(batch > batch_skip)

  print_losses <- function(type){
    tl <- round(data |> filter(type == .env$type) |> pull(loss) |> tail(), 6)
    mn <- min(tl)
    msg <- purrr::map_chr(tl, ~ifelse(.x <= mn, crayon::green(.x), crayon::yellow(.x)))
    cat(type,"\t", msg, '\n')
  }

  epoch = data |> tail(1) |> pull(batch)
  last_lr = data |> tail(1) |> pull(lr)
  
  cat("Epoch:", epoch, "Last learning rate:", last_lr, "\n")
  print_losses("train")
  print_losses("sched")
  if(nrow(data |> filter(type =="eval")) > 0){ print_losses("eval") }
  
  data <- data #|> filter(type != "train")
  # Create the plot with a black theme and log scale for the y-axis
  plot <- ggplot(data, aes(x = batch, y=loss, col=type)) + 
    geom_point(aes(color = type),alpha=0.8, size=2) +
    # geom_smooth(size=1, alpha=0.5, col="red") +
    labs(x = 'Iteration', y = 'Loss', title = 'VAE Losses Over Iterations', color = "Metric") +
    # scale_y_continuous(limits = c(min(data$loss),max(data$loss))) +
    theme_minimal(base_size = 16) +
    theme(
      text = element_text(color = "white"),
      plot.background = element_rect(fill = "black"),
      panel.background = element_rect(fill = "black"),
      legend.background = element_rect(fill = "black", color = "black"),
      legend.position = "bottom",
      legend.title.align = 0.5,
      plot.title = element_text(color = "white"),
      axis.title = element_text(color = "white"),
      axis.text = element_text(color = "white"),
      axis.line = element_line(color = "#533e3e"),
      panel.grid.major = element_line(color = "#533e3e"),
      panel.grid.minor = element_line(color = "#533e3e")
    ) +
    guides(color = guide_legend(title = "Metrics")) + 
    facet_wrap(~type, scales="free_y")

  # Create the directory and save the plot
  dir.create('notebook/plots', recursive = TRUE, showWarnings = FALSE)
  # ggsave('notebook/plots/loss.svg', plot = plot, width = 12, height = 7, dpi = 300)
  ggsave('notebook/plots/loss.png', plot = plot, width = 12, height = 7, dpi = 300)
  return(scheduler_length)
}

last_scheduler_length = 0
while(TRUE){
  last_scheduler_length = draw_plot(last_scheduler_length)
  Sys.sleep(1)
}

File: notebook/loss.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import sys
import time
from termcolor import colored

sns.set(style="darkgrid")

args = sys.argv[1:]
batch_skip = int(args[0]) if len(args) > 0 else 0

def draw_plot(last_scheduler_length=0):
    # Read the first three columns of the metrics file
    data = pd.read_csv('metrics/multitask_loss.tsv', sep='\t', names=['epoch', 'batch', 'type', 'loss', 'lr'], skiprows=1)
    data['type'] = data['type'].replace('scheduler', 'sched')
    scheduler_length = len(data[data['type'] == 'train']['loss'])
    
    if scheduler_length == last_scheduler_length:
        return last_scheduler_length
    
    data['batch'] = data.index + 1
    data = data[data['batch'] > batch_skip]

    def print_losses(loss_type):
        dt = data[data['type'] == loss_type]['loss'].round(6)
        mn = dt.min()
        tl = data[data['type'] == loss_type]['loss'].round(6).tail()
        msg = [colored(x, 'green') if x <= mn else colored(x, 'yellow') for x in tl]
        print(colored(f"{loss_type}\t {' '.join(msg)}", 'white'))
    
    # epoch = data['batch'].iloc[-1]
    # last_lr = data['lr'].iloc[-1]

    # print(colored(f"Epoch: {epoch} Last learning rate: {last_lr}", 'white'))
    print_losses('train')
    print_losses('sched')
    if len(data[data['type'] == 'eval']) > 0:
        print_losses('eval')

    # Create the plot with a black theme and log scale for the y-axis
    g = sns.FacetGrid(data, row="type", hue="type", palette=['#1f77b4', '#ff7f0e', '#2ca02c'], sharex=True, sharey=False, height=5, aspect=2)
    g.map(sns.scatterplot, "batch", "loss", alpha=1.0, s=50, edgecolor=None)
    g.set_titles(row_template="{row_name}", color='white')
    g.set_axis_labels("Iteration", "Loss")
    
    for ax in g.axes.flatten():
        ax.set_facecolor('black')
        ax.figure.set_facecolor('black')
        ax.spines['top'].set_color('white')
        ax.spines['bottom'].set_color('white')
        ax.spines['left'].set_color('white')
        ax.spines['right'].set_color('white')
        ax.tick_params(axis='x', colors='white', which='both')  # Ensure x-axis tick labels and minor ticks are white
        ax.tick_params(axis='y', colors='white', which='both')  # Ensure y-axis tick labels and minor ticks are white
        ax.yaxis.label.set_color('white')
        ax.xaxis.label.set_color('white')
        ax.title.set_color('white')
        ax.grid(True, which='both', axis='both', color='#333333', linestyle='-', linewidth=0.5)
    
    g.add_legend(title='Metric', labelcolor='white', facecolor='black', edgecolor='black')

    # Create the directory and save the plot
    os.makedirs('notebook/plots', exist_ok=True)
    g.savefig('notebook/plots/loss.png', dpi=300, bbox_inches='tight')
    plt.close()

    return scheduler_length

last_run = 0
last_scheduler_length = 0
while True:
    last_scheduler_length = draw_plot(last_scheduler_length)
    time.sleep(15)
